README.md
SayIt
=====

A Django application or project to store transcripts
and present them in a modern, searchable format.

An example of SayIt in action is our website
[sayit.mysociety.org](http://sayit.mysociety.org) which contains a variety of
transcripts, including:

* [the Leveson inquiry](http://leveson.sayit.mysociety.org)
* [the Charles Taylor trial](http://charles-taylor.sayit.mysociety.org)
* [the Plays of Shakespeare](http://shakespeare.sayit.mysociety.org)

SayIt is a [Poplus component](http://poplus.org) 
by [mySociety](http://www.mysociety.org/)

Get involved
------------

If you have transcripts you'd like to be included on our website, or have
the technical skills to create such transcripts, please see
[http://sayit.mysociety.org/about/community](http://sayit.mysociety.org/about/community).

For more information on how to use SayIt in your own Django project or as a
standalone site, please see our documentation site at
[http://mysociety.github.io/sayit/](http://mysociety.github.io/sayit/).

[![Build Status](https://travis-ci.org/mysociety/sayit.png)](https://travis-ci.org/mysociety/sayit)


CHANGES.txt
v2.0, 2019-04-10
    * Django 2 support.
    * Drop support for Django before 1.11.
    * Test fixes for changes on Travis.
    * Add config for ffmpeg binary name.

v1.5, 2017-06-16
    * Django 1.10 and 1.11 support.

v1.4.1, 2016-11-10
    * Fix issue with some exactly-pinned dependencies.
    * Bundle some missing .mo files.

v1.4, 2016-02-24
    * Drop support for Django before 1.8.
    * Add a web interface for importing Akoma Ntoso files. #435
    * Show speaker's speeches most recent first. #482
    * Hebrew and Slovak translations.
    * Improved testing using tox.
    * Use regular arrows in back/forward buttons. #509
    * Akoma Ntoso import
        * Add improved skip/replace/merge import options. #456 #434
        * On directory import, import files in sorted order.
        * Raise exception on empty 'showAs' attributes. #442
        * Warn if docDate missing attribute or doesn't parse. #492
        * Print out statistics in command line import. #460
        * Print traceback for import exceptions. #475
        * Make SSL certificate verification optional. #476
    * Bugfixes
        * Add translation of metadata. #497
        * Fix issue with importing empty docDate/source_url. #484
        * Compare section and speech keys with same type. #472
        * Fix redisplaying of speech date in Django 1.7+.
        * Give sections with no headings a title of 'Untitled'.
        * Fix flake8 part of pre-commit hook, and colour code output.
        * Fix display of speaker name when used in code.

Thanks to James McKinney, Duncan Parkes, Jaroslav Semančík and Matthew
Somerville for their work on this release.

v1.3.2, 2014-12-13
    * Improved initial running of example project, providing login links
      and buttons for adding or importing data. #222 #411 #419

v1.3.1, 2014-12-12
    * Fix error in speaker colour display in Python 3.
    * Akoma Ntoso import
        * Better generic error handling, and of missing speaker TLCPersons

Thanks to Jaroslav Semančík and johnfelipe for reporting these issues.

v1.3, 2014-12-05
    * Delete/ merge speaker front end interface. #407
    * PopIt and Popolo JSON import management command and front end. #231 #310
    * OpenGraph data output. #364 #416
    * Dots in debate subsection timeline, fix section divider display. #372
    * Spanish translation, bundled .mo files. #356
    * Nicer section and speaker forms. #350
    * Update the slug if you edit speaker or section. #433
    * Show more editing buttons on home/ speakers pages. #417
    * Improve speaker sorting, including ability to provide sort name.
    * Use /section instead of /sections for editing URL endpoints.
    * Akoma Ntoso import
        * Options to clobber or skip existing data on import. #387
        * No error if no files found on --dir import. #409
        * Support for source_url. #354
    * Development
        * Quieten unneeded warning. #406
        * flake8 everything, and add flake8 to pre-commit hook. #432
        * Speed up Travis.

Thanks to Duncan Parkes, James McKinney, and Matthew Somerville for their work
on this release.

v1.2.2, 2014-10-01
    * Fix a bug in South migrations introduced in v1.2 (80e3f0b4) where
      migrating could give an error when it tried to rename the content type.
    * Upgrade django-subdomain-instances to get its migrations.
    * Don't show keyboard instructions when no next/previous section.
    * Fix typo in link in README.

Thanks to Andrew Langhorn, Duncan Parkes, Matthew Somerville, and Martin Wright
for their work on this release.

v1.2.1, 2014-09-26
    * Don't error if speaker image caching fails
    * Add new-style migrations for Django 1.7
    * Allow loading of remote Akoma Ntoso files
    * Make it easier to override the ImportCommand behaviour that is
      filesystem-specific
    * Remove requirement to run collectstatic before tests

Thanks to James McKinney and Matthew Somerville for their work on this release.

v1.2, 2014-09-19
    * Cache speaker image from URL, using face recognition for thumbnail
    * Improve display of subsections on section pages
    * Unicode slug support
    * Paginate /speeches
    * Link truncated speech text to speech
    * Tweaks to speech templates, including showing speech titles
    * Add a date picker to speech/ section forms
    * Unhide unimportant form sections which contain errors
    * More translatable strings, including form labels
    * Reduce database queries in get_ancestors and search indexing
    * Model additions:
        * Speech 'type'
        * Section start date and time
        * Section/Speech num, heading (renamed from title) and subheading
    * Akoma Ntoso import
        * Allow anonymous speeches
        * Add support for more preface elements
    * Development
        * Easier local development settings override
        * No requirement to have auth middleware installed

Thanks to James McKinney, Duncan Parkes, Matthew Somerville, Martin Wright, and
Zarino Zappia for their work on this release.

v1.1.1, 2014-07-04
    * Switch to PyPI-hosted fork of django-bleach to remove extra step running
      under Python 3.

v1.1, 2014-07-04
    * Display improvements, including to the default front page, the speech
      detail page, nicer empty content messages and various tweaks
    * Editing improvements, including a much nicer speech editing form, with
      inline creation of new speakers and sections
    * django-popolo is now used to store people data
    * Python 3 compatibility
    * Removal of unneeded dependencies
    * Included git pre-commit hook for CSS compilation
    * Realtime search processer enabled in the example project
    * Akoma Ntoso import
        * Support coverPage and docTitle
        * Log rather than die on error
        * Namespace handling fix

Thanks to Richard Lewis, James McKinney, Duncan Parkes, Matthew Somerville, and
Zarino Zappia for their work on this release.

v1.0, 2014-04-16 -- Initial release.


scrapers/scsl.py
#!/usr/bin/env python

from datetime import datetime

import os
import re

import requests_cache

from scsl.parse import parse_transcript
from utils import *

BASE_DIR = os.path.dirname(__file__)
session_day = requests_cache.core.CachedSession(os.path.join(BASE_DIR, 'data', 'charles-taylor'), expire_after=86400)

class SCSLParser(BaseParser):
    instance = 'charles-taylor'

    def skip_transcript(self, data):
        if data['date'].isoformat() == '2006-07-21': return True # Is garbled
        return False

    def get_transcripts(self):
        old_requests = self.requests
        self.requests = session_day
        transcripts = self.get_url('http://www.sc-sl.org/CASES/ProsecutorvsCharlesTaylor/Transcripts/tabid/160/Default.aspx', 'html')
        self.requests = old_requests
        # Loop through the rows in reverse order (so oldest first)
        for row in transcripts('p'):
            for thing in row.findAll(text=re.compile('\d')):
                text = thing.strip()
                if re.match('\d+$', text):
                    link = thing.find_parent('a')
                    date = date.replace(day=int(text))
                    url = link['href'].replace(' ', '%2B').replace('+', '%2B')

                    if 'tabid/160/www.sc-sl.org' in url:
                        url = re.sub('.*www', 'www', url)
                    if url[0:3] == 'www':
                        url = 'http://%s' % url
                    if url[0] == '/':
                        url = 'http://www.sc-sl.org' + url

                    # Wrong date on index page
                    if date.isoformat() == '2009-06-09':
                        date = date.replace(day=8)

                    yield {
                        'date': date,
                        'url': url,
                        'text': self.get_pdf(url, name=date.isoformat()+'.pdf'),
                    }

                else:
                    date = datetime.strptime(thing, '%B %Y').date()

    def parse_transcript(self, data):
        return parse_transcript(data['text'], data['date'])

    def prettify(self, s):
        s = s.title()
        s = s.replace('Dct-', 'DCT-').replace('Tfi-', 'TF1-').replace('Tf1-', 'TF1-')
        return s

parser = SCSLParser()
parser.run()
